keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -32,218 +31,110 @@ status: provisional
-if err != nil {
-    return err
-}
-waitForPodToBeRunning(ctx, myPod)
-return nil
-```
+We will wrap the API Server's http server and http clients with [othttp](https://github.com/open-telemetry/opentelemetry-go/tree/master/plugin/othttp) to get spans for incoming and outgoing http requests, and add the [otgrpc](https://github.com/grpc-ecosystem/grpc-opentracing/tree/master/go/otgrpc) DialOption to the etcd grpc client.  This generates spans for all sampled incoming requests and propagates context with all client requests.
 
-A previous iteration of this proposal suggested controllers should export a "Root Span" when ending a trace (described in [Controller Behavior](#controller-behavior) above).  However, that would limit a trace to being associated with a single object, since a "Root Span" defines the scope of the trace.  More generally, we shouldn't assume that the creation or update of a single object represents the entirety of end-user intent.  The user or system using kubernetes determines what the user intent is, not kubernetes controllers.
+Note that some clients of the API Server, such as webhooks, may make reentrant calls to the API Server.  To gain the full benefit of tracing, such clients should propagate context with requests back to the API Server.
 
-Tracing in a kubernetes cluster must be a composable component within a larger system, and allow external users or systems to define the "Root Span" that defines and bounds the scope of a trace.
+### Kubectl integration
 
-### In-tree changes
+Fundamentally, a trace is a graph showing the fulfilment of user intent.  To make tracing in kubernetes useful, we need to ensure users have the ability to declare their intent for their request to be traced.  Add a new `--trace` argument to `kubectl`, which generates a new trace context, sets the trace context to be sampled, and uses the context when sending requests to the API Server.  The option is false by default.
 
-#### Vendor the Tracing Framework
+### Vendor OpenTelemetry and the OT Exporter
 
 This KEP proposes the use of the [OpenTelemetry tracing framework](https://opentelemetry.io/) to create and export spans to configured backends.
 
-While in alpha, controllers should use the OpenTelemetry exporter, which exports traces to the [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector). The OpenTelemetry collector allows importing and configuring exporters for trace storage backends to be done out-of-tree in addition to other useful features.
+The API Server will use the [OpenTelemetry exporter format](https://github.com/open-telemetry/opentelemetry-proto), which exports traces to a local port.  This format is compatible with the [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector), which allows importing and configuring exporters for trace storage backends to be done out-of-tree in addition to other useful features.
```

Comment:
We need some more detail-- are these going to be synchronous RPCs? On every request? Batched and sent afterwards? Something else?

Reply:
These are batched RPCs sent afterwards.  OpenTelemetry has the notion of "processors" 
 for spans.  One of the built-in ones is the [batching processor](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#batching-processor).